jscomp/bsb/bsb_helper_dep_graph.ml

```diff
@@ -0,0 +1,114 @@
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. *)
+
+let sort_files_by_dependencies ~domain dependency_graph =
+  let next current =
+    (String_map.find_exn  current dependency_graph) in
+  let worklist = ref domain in
+  let result = Queue.create () in
+  let rec visit visiting path current =
+    if String_set.mem current visiting then
+      Format.fprintf Format.err_formatter "Cyclic depends : @[%a@]"
+        (Format.pp_print_list ~pp_sep:Format.pp_print_space
+           Format.pp_print_string)
+        (current::path)
+    else if String_set.mem current !worklist then
+      begin
+        next current |>
+        String_set.iter
+          (fun node ->
+             if  String_map.mem node  dependency_graph then
+               visit (String_set.add current visiting) (current::path) node)
+        ;
+        worklist := String_set.remove  current !worklist;
+        Queue.push current result ;
+      end in
+  while not (String_set.is_empty !worklist) do
+    visit String_set.empty []  (String_set.choose !worklist)
+  done;
+  if Js_config.get_diagnose () then
+    Format.fprintf Format.err_formatter
+      "Order: @[%a@]@."
+      (Ext_format.pp_print_queue
+         ~pp_sep:Format.pp_print_space
+         Format.pp_print_string)
+      result ;
+  result
+;;
+
+let simple_collect_from_main ?alias_map ast_table main_module =
+  let visited = String_hashtbl.create 31 in
+  let result = Queue.create () in
+  let next module_name : String_set.t =
+    let module_set =
+      match String_map.find_exn module_name ast_table with
+      | exception _ -> String_set.empty
+      | x -> x
+    in
+    match alias_map with
+    | None -> module_set
+    | Some map ->
+      String_set.fold (fun x acc -> String_set.add (String_hashtbl.find_default map x x) acc  ) module_set String_set.empty
+  in
+  let rec visit visiting path filepath =
+    let current = Ext_filename.module_name_of_file_if_any filepath in
+    if String_set.mem current visiting  then
+      Format.fprintf Format.err_formatter "Cyclic depends : @[%a@]"
+        (Format.pp_print_list ~pp_sep:Format.pp_print_space
+           Format.pp_print_string)
+        (current::path)
+    else
+    if not (String_hashtbl.mem visited current)
+    && String_map.mem current ast_table then
+      begin
+        String_set.iter
+          (visit
+             (String_set.add current visiting)
+             (current::path))
+          (next current) ;
+        Queue.push filepath result;
+        String_hashtbl.add visited current ();
+      end in
+  visit (String_set.empty) [] main_module ;
+  result
+
+let get_otherlibs_dependencies dependency_graph file_extension =
+  let set_of_otherlib_deps = String_map.fold (fun k v acc ->
+    if String_set.mem "Unix" v then
+      String_set.add ("unix" ^ file_extension) acc
+    else if String_set.mem "Bigarray" v then
```

Comment:
I'm not totally sure how you want this refactor. I think a constant number of Set lookups might be the fastest way to do this.